Rename ImmutableArray to ArraySeq

lrytz opened the pull request:
Fixes https://github.com/scala/bug/issues/10836

lrytz committed:
Rename ImmutableArray to ArraySeq

NthPortal commented:
I think we may need another collection - a `ConSeq` - for this

lrytz commented:
rebased on the insource-partest pr for now to test, but unfortunately that created 100 jobs on jenkins, as there's no `[ci: last-only]` here...

lrytz committed:
Rename WrappedArray to ArraySeq

lrytz commented:
Tests look good, but the `specialized` category failed because I need to build a new `instrumented.jar`. I did that recently, so no big deal.

lrytz commented:
Added another commit that renames `WrappedArray` to `ArraySeq`. So basically there's `mutable.ArraySeq` and `immutable.ArraySeq`. @szeiger suggested this in https://github.com/scala/bug/issues/10836. I'm not 100% sure it's a good idea, because creating a `mutable.ArraySeq` is not always explicit, due to `Predef.wrapXArray`. It might be too confusing. Let's discuss!

adriaanm commented:
/synch

adriaanm commented:
/rebuild

sjrd commented on src/library/scala/collection/mutable/package.scala:
```diff
@@ -2,10 +2,6 @@ package scala.collection
-  @deprecated("Use WrappedArray instead of ArraySeq; it can represent both, boxed and unboxed arrays", "2.13.0")
-  type ArraySeq[X] = WrappedArray[X]
-  @deprecated("Use WrappedArray instead of ArraySeq; it can represent both, boxed and unboxed arrays", "2.13.0")
-  val ArraySeq = WrappedArray.untagged
```
I think we need the deprecated aliases in the other direction, if we're doing this change.

lrytz replied:
Yes, thanks for pointing that out. Though I think we should drop the second commit, keep `mutable.WrappedArray`. Waiting for opinions on that still :-)

julienrf commented:
> Users of 2.12 `ArraySeq` would have to use `ArraySeq.untagged`. We could add a migration warning.

I agree. And we can easily define a rewrite rule replacing `ArraySeq.empty[X]` to `ArraySeq.untagged.empty[X]`, and same for all the other operations that now take an implicit `ClassTag` at

lrytz commented:
It turns out just rebuilding scalacheck with a local Scala version is not enough, because the scalacheck artifact depends on the Scala library it was built with. So that Scala version needs to be on maven... We can either
  1. Release another `2.13.0-M4-pre-abcd123` to maven (with the rename) and the build scalacheck using that
  2. In-source scalacheck
  3. other ideas?

I vote for 2.

lrytz committed:
In-source scalacheck

Sources from https://github.com/lrytz/scalacheck/commit/1fd5eef92ca8775cb689ead0d10b8798aebd5485

lrytz commented:
Added a commit that in-sources scalacheck. TODO: update IntelliJ templates.

lrytz committed:
Don't run macroAnnot tests for now

We get some linkage errors now. sbt seems to compile the compiler
interface against the current version, which is 2.13.0-pre-SNAPSHOT,
so something arbitrary. The scalaInstance in the macroAnnot project
is the quick instance. That's probably the cause.

We should convert macroAnnot tests to partest, where we do this
bootstrap.

szeiger commented on src/library/scala/Array.scala:
```diff
@@ -494,11 +494,11 @@ object Array {
     // in pattern matcher.  @PP: I noted in #4364 I think the behavior is correct.
-    // Is ImmutableArray safe here? In 2.12 we used to call .toIndexedSeq which copied the array
-    // instead of wrapping it in a WrappedArray but it appears unnecessary.
+    // Is ArraySeq safe here? In 2.12 we used to call .toIndexedSeq which copied the array
```
It turns out it is not. We need to make a defensive copy. Varargs use `immutable.Seq` so we can't just return a `mutable.ArraySeq` but the Seq can leak out of a pattern match.

sjrd commented:
FYI, it turns out Scala.js also depends on this PR to be merged, if it does and what portions of it do :-s I had not realized that before.

I've already updated after #6620.

lrytz commented:
I'll merge this as soon as it goes green.
